### ISSUE TITLE
Version 1.2.0 uses a breaking version of UglifyJS (2.4.19)

### ISSUE THREAD

**Issue by mpetrovich** [terinjokes/gulp-uglify#98]
[Version 1.2.0 updated UglifyJS to v2.4.19](https://github.com/terinjokes/gulp-uglify/commit/4ead3aaefb3abc7f95997ee87312baf36a07b7f3). However, [UglifyJS v2.4.18 introduced a breaking change](https://github.com/mishoo/UglifyJS2/issues/673) that breaks one or more of our vendor scripts.

As a result, I've had to lock `gulp-uglify` to version 1.1.0 for now.

**Comment by mpetrovich** [comment]
[UglifyJS just released a patch](https://github.com/mishoo/UglifyJS2/issues/673#issuecomment-92411752) that addresses this issue.

**Comment by terinjokes** [comment]
You can `require('gulp-uglify/minifier')` and pass in your own version of UglifyJS and options now.

```javascript
var minifer = require('gulp-uglfiy/minifer');
var uglifyjs = require('uglify-js');

gulp.src('src/*.js')
  .pipe(minifier({}, uglifyjs)
  .pipe(gulp.dest('dist/'));
```

This isn't documented yet, but I'm pretty happy with it so far, so I suspect it will stay.

**Comment by mpetrovich** [comment]
Good idea @terinjokes, thank you!

**Comment by terinjokes** [comment]
@mpetrovich if you have a combination that works, lmk, and I can update the package.json.

**Comment by tajo** [comment]
```
/Users/miksu/Projects/flow/node_modules/gulp-uglify/node_modules/deap/lib/deap.js:56
		Object.keys(b).forEach(function(p) {
		       ^
TypeError: Object.keys called on non-object
    at Function.keys (native)
    at /Users/miksu/Projects/flow/node_modules/gulp-uglify/node_modules/deap/lib/deap.js:56:10
    at Array.forEach (native)
    at deepExtend (/Users/miksu/Projects/flow/node_modules/gulp-uglify/node_modules/deap/lib/deap.js:55:27)
    at setup (/Users/miksu/Projects/flow/node_modules/gulp-uglify/minifier.js:18:17)
    at DestroyableTransform.minify [as _transform] (/Users/miksu/Projects/flow/node_modules/gulp-uglify/minifier.js:55:19)
    at DestroyableTransform.Transform._read (/Users/miksu/Projects/flow/node_modules/gulp-uglify/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:184:10)
    at DestroyableTransform.Transform._write (/Users/miksu/Projects/flow/node_modules/gulp-uglify/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:172:12)
    at doWrite (/Users/miksu/Projects/flow/node_modules/gulp-uglify/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:237:10)
    at writeOrBuffer (/Users/miksu/Projects/flow/node_modules/gulp-uglify/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:227:5)
```
I guess it's related. I got this after update to 1.2.0.

**Comment by terinjokes** [comment]
@tajo can you show the task from your gulpfile?

**Comment by tajo** [comment]
@terinjokes There you go:
```javascript
'use strict';
var gulp = require('gulp');
var uglify = require('gulp-uglify');
var browserify = require('browserify');
var streamify = require('gulp-streamify');
var babelify = require('babelify');
var stripDebug = require('gulp-strip-debug');

gulp.task('build', function() {
  process.env.NODE_ENV = 'production';
  browserify({
    entries: [paths.ENTRY_POINT],
    transform: [babelify]
  })
    .bundle()
    .pipe(source(paths.OUT))
    .pipe(streamify(uglify(paths.OUT)))
    .pipe(streamify(stripDebug()))
    .pipe(gulp.dest(paths.DEST));
});
```

**Comment by Dudemullet** [comment]
@tajo I'm getting the same issue, have you found a workaround or fix for it yet?

**Comment by tajo** [comment]
@Dudemullet Unfortunately not. I tried

> require('gulp-uglify/minifier') and pass in your own version of UglifyJS and options now

and it did not work. I didn't have much time to investigate, so I just downgraded to `1.1.0`

**Comment by Dudemullet** [comment]
Thanks @tajo, guess I'm downgrading in the meantime too.

FWIW, I also tried

> require('gulp-uglify/minifier') and pass in your own version of UglifyJS and options now

And that didn't work for me either.

**Comment by terinjokes** [comment]
@tajo @Dudemullet this is exactly how `gulp-uglify` loads in UglifyJS (see the very short and simple index.js: https://github.com/terinjokes/gulp-uglify/blob/ee1798746f2b056be9eea419dea97fedcbe93330/index.js

@tajo I believe I have your Object.keys issue fixed on the branch for 1.3.0.

**Comment by terinjokes** [comment]
@mpetrovich I've upgraded the 1.3.0 branch to 2.4.21, hopefully that's good enough.

**Comment by holic** [comment]
Curious on the decision to tie this module to a specific version of UglifyJS instead of taking advantage of semver?

**Comment by terinjokes** [comment]
This module used to use SemVer for UglifyJS but found that it was unstable
and I would get issues for suddenly breaking people's builds. People would
roll back to previous versions of gulp-uglify and find that it still
doesn't work. It's the same reason why I do minor bumps for UglifyJS
version changes.

A future major release will finalize making UglifyJS injectable, and I'll
likely switch UglifyJS back to SemVer.

**Comment by holic** [comment]
That's understandable, but I think both solutions are less than ideal. I think [peer dependencies](http://blog.nodejs.org/2013/02/07/peer-dependencies/) is the solution you're looking for. It would allow me to specify in my dependencies the version of UglifyJS and gulp-uglify without explicitly logging gulp-uglify to a version.

So the install process would instead become:
```
npm install --save-dev gulp-uglify uglify-js
```

Happy to open a PR with this proposed change.

**Comment by terinjokes** [comment]
You can use the dependency injection pattern if you want to use your own
UglifyJS.

**Comment by sochka** [comment]
I'm sometimes randomly getting this error:

```
/app/node_modules/gulp-uglify/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:972
  for (var i = 0, l = xs.length; i < l; i++) {
                        ^
TypeError: Cannot read property 'length' of null
    at forEach (/app/node_modules/gulp-uglify/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:972:25)
    at flow (/app/node_modules/gulp-uglify/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:634:7)
    at DestroyableTransform.<anonymous> (/app/node_modules/gulp-uglify/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:613:7)
    at emitNone (events.js:67:13)
    at DestroyableTransform.emit (events.js:163:7)
    at onwriteDrain (/app/node_modules/gulp-size/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:300:12)
    at afterWrite (/app/node_modules/gulp-size/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:288:5)
    at onwrite (/app/node_modules/gulp-size/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:281:7)
    at WritableState.onwrite (/app/node_modules/gulp-size/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:106:5)
    at afterTransform (/app/node_modules/gulp-size/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:104:5)
    at TransformState.afterTransform (/app/node_modules/gulp-size/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:79:12)
```

Tested on iojs.

**Comment by sloppylopez** [comment]
Hi, I am surprised some people say it is working for them with gulp-uglify 1.1.0
If you clone this project to use it as an example:
https://github.com/paislee/healthy-gulp-angular

and you change the package.json from     "gulp-uglify": "^1.1.0",
to
    "gulp-uglify": "1.1.0",

and do 
bower i && npm i

If you execute 
gulp

everythign works, including the uglifycation of the vendor scripts

However, if you do

bower i --save textAngular

to put an extra dependency in bower

If you do next gulp

you get this:

events.js:85
      throw er; // Unhandled 'error' event
            ^
Error
    at new JS_Parse_Error (/home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:189:18)
    at js_error (/home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:197:11)
    at parse_error (/home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:299:55)
    at next_token (/home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:544:9)
    at /home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:415:16
    at /home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:509:24
    at handle_slash (/home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:485:20)
    at Object.next_token [as input] (/home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:538:27)
    at next (/home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:643:25)
    at semicolon (/home/sloppylopez/work/healthy-gulp-angular/node_modules/gulp-uglify/node_modules/uglify-js/lib/parse.js:690:30)

**Comment by terinjokes** [comment]
@sloppylopez Please upgrade to 1.3.0 and attach an error handler to gulp-uglify so you can see the error message and line number from UglifyJS.

**Comment by sloppylopez** [comment]
Thanks, will do asap

**Comment by sloppylopez** [comment]
Aw, at first sight following my described method now it seems to work, i am double checking that everything is tight and i did not somethign wrong, BUT IT REALLY LOOKS LIKE ITS WORKING, THANKS SO MUCH

**Comment by sloppylopez** [comment]
Yes, it works with 1.3.0
Thanks I owe you 1

**Comment by cwbuecheler** [comment]
Using gulp-uglify 1.4.0 I get the same error. If I downgrade to 1.1.0 it works, but I figured it was better to report the error. Let me know if I screwed something up!

Code:
```javascript
// Build Task for Production
gulp.task('build', function() {
  gulp.src(path.JS)
    .pipe(react())
    .pipe(concat(path.MINIFIED_OUT))
    .pipe(uglify(path.MINIFIED_OUT))
    .pipe(gulp.dest(path.DEST_BUILD));
});
```

Error:
```
D:\sites\node\reacttest3\node_modules\gulp-uglify\node_modules\deap\lib\deap.js:56
                Object.keys(b).forEach(function(p) {
                       ^
TypeError: Object.keys called on non-object
    at Function.keys (native)
    at D:\sites\node\reacttest3\node_modules\gulp-uglify\node_modules\deap\lib\deap.js:56:10
    at Array.forEach (native)
    at deepExtend (D:\sites\node\reacttest3\node_modules\gulp-uglify\node_modules\deap\lib\deap.js:55:27)
    at setup (D:\sites\node\reacttest3\node_modules\gulp-uglify\minifier.js:19:17)
    at DestroyableTransform.minify [as _transform] (D:\sites\node\reacttest3\node_modules\gulp-uglify\minifier.js:58:19)
    at DestroyableTransform.Transform._read (D:\sites\node\reacttest3\node_modul
es\gulp-uglify\node_modules\through2\node_modules\readable-stream\lib\_stream_tr
ansform.js:172:10)
    at DestroyableTransform.Transform._write (D:\sites\node\reacttest3\node_modules\gulp-uglify\node_modules\through2\node_modules\readable-stream\lib\_stream_transform.js:160:12)
    at doWrite (D:\sites\node\reacttest3\node_modules\gulp-uglify\node_modules\through2\node_modules\readable-stream\lib\_stream_writable.js:326:12)
    at writeOrBuffer (D:\sites\node\reacttest3\node_modules\gulp-uglify\node_modules\through2\node_modules\readable-stream\lib\_stream_writable.js:312:5)
```

**Comment by terinjokes** [comment]
You can't pass a string to gulp-uglify.

**Comment by cwbuecheler** [comment]
Oh. Is that new? It seems like a lot of the information I've been seeing does pass a string, and passing a string works in 1.1.0. Anyway, it works if I just don't pass it the string, so thanks!

**Comment by terinjokes** [comment]
gulp-uglify has never done anything meaningful with a string, the options
argument has always been an object.

**Comment by cwbuecheler** [comment]
Got it. Thanks for the clarification! Unsure why the tutorials I was reading were passing it a string.

**Comment by terinjokes** [comment]
If you want to link me to some of them, I'm happy to email them to try to
correct the problems.

-- 
#Terin Stock

**Comment by cwbuecheler** [comment]
I just checked and for the main tutorial I was using, people in the comments are already pointing this out, so I suspect it'll be corrected soon. It was this one, for the record: http://tylermcginnis.com/reactjs-tutorial-pt-2-building-react-applications-with-gulp-and-browserify/

I thought I'd seen it elsewhere but can't find any other tutorials that indicate it should take a string, so I must've been mistaken.

**Comment by terinjokes** [comment]
Fixed in 1.4.1. Please open new issues for any remaining other issues.
